compiler/luci/service/src/CircleShapeInferenceHelper.cpp

```diff
@@ -16,6 +16,104 @@
 
 #include "CircleShapeInferenceHelper.h"
 
+#include <oops/InternalExn.h>
+
+using namespace luci::sinf;
+
+namespace
+{
+
+/**
+ * @brief  Expand shape x and y to same rank by align right and filling with 1
+ */
+void expand_rank(loco::TensorShape &x, loco::TensorShape &y)
+{
+  auto x_rank = x.rank();
+  auto y_rank = y.rank();
+
+  if (x_rank == y_rank)
+    return;
+
+  TensorShapeExpander x_exp(x);
+  TensorShapeExpander y_exp(y);
+
+  auto xy_rank = std::max(x_rank, y_rank);
+
+  x = x_rank > y_rank ? x : x_exp.to(xy_rank);
+  y = y_rank > x_rank ? y : y_exp.to(xy_rank);
+}
+
+/**
+ * @brief  Return shape of expanded dimension of input x and y having same rank
+ */
+loco::TensorShape expand_dimension(const loco::TensorShape &x, const loco::TensorShape &y)
+{
+  assert(x.rank() == y.rank()); // FIX_CALLER_UNLESS
+
+  auto rank = x.rank();
+
+  loco::TensorShape output_shape;
+
+  output_shape.rank(rank);
+
+  // Shape inference rule (commutative)
+  // Values of x, y, and Result are categorized as 1, X, ? where
+  // X is a positive integer greater than 1 and ? is unknown.
```

Comment:
Not sure 1, X, ? are rank or dimension values. It would be nice to have some comments for unit of this values.

Reply:
(optional) how about using `N` instead of `X` ?
first look of `X` gave me meaning of `invalid`

Reply:
I've updated the comment to specify they are dimension values.

I've also changed the code to use `N`.